### PR TITLE
Feature: allow passing requestParameters to s3 integration

### PR DESCRIPTION
**Pull request by erezrokah**
Should resolve the discussion here https://github.com/horike37/serverless-apigateway-service-proxy/pull/36

### CODE REVIEW

**Comment by horike37** [comment]
Nice approach 👍 Thanks @erezrokah 

@blair55
Would this make your problem resolve?

**Comment by blair55** [comment]
Yes it would resolve my problem! :-D

In general, using folders to organise S3 bucket contents is preferable, hence the intention to use the current time. But this is a great solution in the meantime!

Thank you @erezrokah @horike37 @theburningmonk

File: README.md [hunk 1]
```diff
@@ -176,6 +176,27 @@ Sample request after deploying.
...
+        method: post
+        action: PutObject
+        bucket:
+          Ref: S3Bucket
+        key: static-key.json
+        cors: true
+
+        requestParameters:
+          # use auto generated object key (will override the static key)
+          'integration.request.path.object': 'context.requestId'
```
Comment (theburningmonk) [inline README.md]: I'm confused by the configuration here, so we allow a `key` as well as `requestParameter` and let the latter override the first? why not making them mutually exclusive?

Comment (theburningmonk) [reply README.md]: do you know if we can have other request parameters? maybe for specifying S3 metadata? can they be specified via request params?

Comment (erezrokah) [reply README.md]: > I'm confused by the configuration here, so we allow a `key` as well as `requestParameter` and let the latter override the first? why not making them mutually exclusive?

You're right, this is a bit confusing. I thought about making them mutually exclusive, but it requires specifically checking for the 'integration.request.path.object' property in the Joi schema. Let me have another look if I can do it in a non-ugly way

Comment (erezrokah) [reply README.md]: > do you know if we can have other request parameters? maybe for specifying S3 metadata? can they be specified via request params?

I'll test if I can set the cache control for example per https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html

Comment (theburningmonk) [reply README.md]: thanks, much appreciated. I like the approach, just not keen on having to put in a redundant `key` configuration only to then override it with another thing.

**Comment by erezrokah** [comment]
Made `key` and `requestParameters` mutually exclusive when setting the `integration.request.path.object` property and added an example on how to set an additional header.

File: README.md [hunk 2]
```diff
@@ -176,6 +176,27 @@ Sample request after deploying.
...
+    - s3:
+        path: /s3
+        method: post
+        action: PutObject
+        bucket:
+          Ref: S3Bucket
+        cors: true
+
+        requestParameters:
+          # if requestParameters has a 'integration.request.path.object' property you should remove the key setting
```
Comment (theburningmonk) [inline README.md]: 👍

File: lib/apiGateway/schema.js [hunk 3]
```diff
@@ -157,7 +157,15 @@ const proxiesSchemas = {
         .valid('GetObject', 'PutObject', 'DeleteObject')
         .required(),
       bucket: stringOrRef.required(),
-      key: key.required()
+      // don't accept a key when requestParameters has a 'integration.request.path.object' property
+      key: Joi.when('requestParameters', {
```
Comment (theburningmonk) [inline lib/apiGateway/schema.js]: nice!

**Comment by blair55** [comment]
Early tests show its working as expected 👍 

Thanks again all!

**Pull request merged**